File: Python x EDAII/Grafos/Boruvka.py
from Grafo import Grafo
from Arista import Arista
from Vertice import Vertice

class Boruvka:

	def __init__(self):
		'Constructor de la clase'
		pass

	def costo(self,a):
		'Permite ordenar la cola de prioridad'
		return a.costo

	def boruvka(self,g):
		componentes = []
		aristasVisitadas = []
		aristasTotales = []

		# Generar los componentes que contendrán sus respectivos vertices y aristas
		for i in range(0,len(g.vertices)):
			c = Grafo(False,False)
			componentes.append(c)

			s = g.vertices[i]
			componentes[i].vertices.append(s)

			pq = []	# Cola de prioridad

			# Añade todas las aristas adyacentes al vertice actual a la cola de prioridad
			for j in range(0,len(s.aristas)):
				pq.append(s.aristas[j])
			pq.sort(key=self.costo)	# Ordena la cola de prioridad

			# Añade las aristas adyacentes a su respectivo componente
			for j in range(0,len(s.aristas)):
				ar = pq[0]
				del pq[0]
				componentes[i].aristas.append(ar)

			# Remueve la arista de menor valor de ese vertice actual
			minima = componentes[i].aristas[0]

			# Verificar que las aristas no se dupliquen
			if self.buscarArista(aristasVisitadas, minima) == False:
				aristasVisitadas.append(minima)
				aristasTotales.append(minima)

		# Primera Iteración: conectar los vértices en componentes
		componentes.clear()

		while len(aristasVisitadas) > 0:
			c = Grafo(False, False)
			componentes.append(c)
			ind = len(componentes)-1

			ruta = []
			indicesTotales = []
			a = aristasVisitadas[0]
			del aristasVisitadas[0]

			ruta.append(a.v)
			ruta.append(a.u)
			componentes[ind].aristas.append(a)

			self.buscarVerticeIzq(a.v, aristasVisitadas, ruta, indicesTotales)
			self.buscarVerticeDer(a.u, aristasVisitadas, ruta, indicesTotales)

			# Ordenar los indices para remover las aristas de la cola en determinado orden
			indicesTotales.sort()

			# Agrega las aristas que conectan a los vértices de un componente
			for i in range(len(indicesTotales)-1,-1,-1):
				index = indicesTotales[i]
				ar = aristasVisitadas[index]
				componentes[ind].aristas.append(ar)
				del aristasVisitadas[index]

			# Asigna la ruta de vertices generada para cada componente
			componentes[ind].vertices = ruta

		# Imprimir el contenido de cada componente generado
		for i in range(0,len(componentes)):
			print("Componente: ",i)

			for j in range(0,len(componentes[i].vertices)):
				print(componentes[i].vertices[j].nombre,", ")
			print(" ")

			for j in range(0,len(componentes[i].aristas)):
				componentes[i].aristas[j].imprimirArista()
			print(" ")

		# Segunda y Final Iteración: conectar las componentes en una sola (árbol Mínimo)

		# Asignar las aristas minimas que conecten a cada nuevo componente
		for i in range(0,len(componentes)):
			pq = []

			for j in range(0,len(componentes[i].vertices)):
				for k in range(0,len(componentes[i].vertices[j].aristas)):
					act = componentes[i].vertices[j].aristas[k]

					# Si la arista adyacente al vertice no forma parte del componente, se agrega para
					# analizar conexiones entre componentes
					if self.buscarArista(componentes[i].aristas, act) == False:
						pq.append(act)

			# Una vez analizado todo el componente se procede a elegir la arista mínima
			pq.sort(key=self.costo) # Ordena la cola de prioridad
			minima = pq[0]
			del pq[0]

			# Verificar que las aristas no se dupliquen
			if self.buscarArista(aristasVisitadas, minima) == False:
				aristasVisitadas.append(minima)
				aristasTotales.append(minima)

		return aristasTotales

	def buscarArista(self,aristasVisitadas,a):
		'Busca una arista almacenada en una lista'
		encontrado = False
		for i in range(0,len(aristasVisitadas)):
			aux = aristasVisitadas[i]

			if (aux.v == a.v and aux.u == a.u) or (aux.v == a.u and aux.u == a.v):
				encontrado = True

		return encontrado

	def buscarVerticeIzq(self,actual,aristasVisitadas,ruta,indicesTotales):
		'Busca vertices conectados entre sí para añadirlos a un componente (hacia la izquierda)'
		indices = []

		# Busca los vertices que conectan por la derecha
		for i in range(0,len(aristasVisitadas)):
			if aristasVisitadas[i].u == actual:
				indices.append(i)
				indicesTotales.append(i)

		# Añade los vertices a la ruta sin que se repitan
		for i in range(0,len(indices)):
			aux = aristasVisitadas[indices[i]].v
			self.añadirVertice(aux,ruta)

		# Búsqueda recursiva
		for i in range(0,len(indices)):
			self.buscarVerticeIzq(aristasVisitadas[indices[i]].v, aristasVisitadas, ruta, indicesTotales)

	def buscarVerticeDer(self,actual,aristasVisitadas,ruta,indicesTotales):
		'Busca vertices conectados entre sí para añadirlos a un componente (hacia la derecha)'
		indices = []

		# Busca los vertices que conectan por la izquierda
		for i in range(0,len(aristasVisitadas)):
			if aristasVisitadas[i].v == actual:
				indices.append(i)
				indicesTotales.append(i)

		# Añade los vertices a la ruta sin que se repitan
		for i in range(0,len(indices)):
			aux = aristasVisitadas[indices[i]].u
			self.añadirVertice(aux,ruta)

		# Búsqueda recursiva
		for i in range(0,len(indices)):
			self.buscarVerticeDer(aristasVisitadas[indices[i]].u, aristasVisitadas, ruta, indicesTotales)

	def añadirVertice(self,v,ruta):
		'Añade un vertice sin que se repita en una lista'
		repetido = False
		for i in range(0,len(ruta)):
			if ruta[i] == v:
				repetido = True

		# Si no se repite el vertice, se inserta
		if repetido == False:
			ruta.append(v)

# Ejecutando los objetos
programa = Boruvka()
g = Grafo(False,True);

# Generando nodos
va = Vertice("a");
vb = Vertice("b");
vc = Vertice("c");
vd = Vertice("d");
ve = Vertice("e");
vf = Vertice("f");
vi = Vertice("i");
vj = Vertice("j");

# Añadiendo nodos al grafo
g.añadirVertice(va);
g.añadirVertice(vb);
g.añadirVertice(vc);
g.añadirVertice(vd);
g.añadirVertice(ve);
g.añadirVertice(vf);
g.añadirVertice(vi);
g.añadirVertice(vj);

aa = Arista(va,vb,20)
ab = Arista(va,vf,50)
ac = Arista(va,vi,10)
ad = Arista(vi,vj,40)
ae = Arista(vb,vc,5)
af = Arista(vf,vj,30)
ag = Arista(vi,vd,8)
ah = Arista(vd,ve,4)
ai = Arista(vc,ve,15)

# Conectando nodos
g.añadirArista(aa,20)
g.añadirArista(ab,50)
g.añadirArista(ac,10)
g.añadirArista(ad,40)
g.añadirArista(ae,5)
g.añadirArista(af,30)
g.añadirArista(ag,8)
g.añadirArista(ah,4)
g.añadirArista(ai,15)

g.imprimirGrafo()

arbolMin = programa.boruvka(g)

print("Árbol de Expansión Mínimo Boruvka")
for i in range(0,len(arbolMin)):
	arbolMin[i].imprimirArista()

File: EDAII x Java/Grafos/Boruvka.java
import java.util.List;
import java.util.LinkedList;
import java.util.Comparator;
import java.util.PriorityQueue;

public class Boruvka{

	public static void main(String[] args){
		Grafo g = new Grafo(false,true);

		// Generando nodos
		Vertice va = new Vertice("a");
		Vertice vb = new Vertice("b");
		Vertice vc = new Vertice("c");
		Vertice vd = new Vertice("d");
		Vertice ve = new Vertice("e");
		Vertice vf = new Vertice("f");
		Vertice vi = new Vertice("i");
		Vertice vj = new Vertice("j");

		// Añadiendo nodos al grado
		g.añadirVertice(va);
		g.añadirVertice(vb);
		g.añadirVertice(vc);
		g.añadirVertice(vd);
		g.añadirVertice(ve);
		g.añadirVertice(vf);
		g.añadirVertice(vi);
		g.añadirVertice(vj);

		// Generando aristas
		Arista aa = new Arista(va,vb);
		Arista ab = new Arista(va,vf);
		Arista ac = new Arista(va,vi);
		Arista ad = new Arista(vi,vj);
		Arista ae = new Arista(vb,vc);
		Arista af = new Arista(vf,vj);
		Arista ag = new Arista(vi,vd);
		Arista ah = new Arista(vd,ve);
		Arista ai = new Arista(vc,ve);

		// Conectando nodos
		g.añadirArista(aa,20);
		g.añadirArista(ab,50);
		g.añadirArista(ac,10);
		g.añadirArista(ad,40);
		g.añadirArista(ae,5);
		g.añadirArista(af,30);
		g.añadirArista(ag,8);
		g.añadirArista(ah,4);
		g.añadirArista(ai,15);

		g.imprimirGrafo();

		LinkedList<Arista> arbolMin = boruvka(g);

		System.out.println("Árbol de Expansión Mínimo Boruvka");
		for (int i=0; i<arbolMin.size(); i++){
			arbolMin.get(i).imprimirArista();
		}
		System.out.println("");
	}

	public static LinkedList<Arista> boruvka(Grafo g){
		LinkedList<Grafo> componentes = new LinkedList<>();
		LinkedList<Arista> aristasVisitadas = new LinkedList<>();
		LinkedList<Arista> aristasTotales = new LinkedList<>();

		// Generar los componentes que contendrán sus respectivos vertices y aristas
		for (int i=0; i<g.vertices.size(); i++){
			componentes.add(new Grafo());

			Vertice s = g.vertices.get(i);
			componentes.get(i).vertices.add(s);

			Comparator<Arista> comparador = new ComparadorArista(); // Ordena la cola de prioridad
			PriorityQueue<Arista> pq = new PriorityQueue<>(comparador);
			
			// Añade todas las aristas adyacentes al vertice actual a la cola de prioridad
			for (int j=0; j<s.aristas.size(); j++){
				pq.add(s.aristas.get(j));
			}

			// Añade las aristas adyacentes a su respectivo componente
			for (int j=0; j<s.aristas.size(); j++){
				componentes.get(i).aristas.add(pq.poll());
			}

			// Remueve la arista de menor valor de ese vertice actual
			Arista minima = componentes.get(i).aristas.get(0);

			// Verificar que las aristas no se dupliquen
			if (buscarArista(aristasVisitadas,minima) == false){
				aristasVisitadas.add(minima);
				aristasTotales.add(minima);
			}
		}

		// Primera Iteración: conectar los vértices en componentes
		componentes = new LinkedList<>();

		while (aristasVisitadas.size() > 0){
			componentes.add(new Grafo());
			int ind = componentes.size()-1;

			LinkedList<Vertice> ruta = new LinkedList<>();
			LinkedList<Integer> indicesTotales = new LinkedList<>();
			Arista a = aristasVisitadas.poll();

			ruta.add(a.v);
			ruta.add(a.u);
			componentes.get(ind).aristas.add(a);

			buscarVerticeIzq(a.v,aristasVisitadas,ruta,indicesTotales);
			buscarVerticeDer(a.u,aristasVisitadas,ruta,indicesTotales);

			// Ordenar los indices para remover las aristas de la cola en determinado orden
			bubbleSort(indicesTotales);
			
			// Agrega las aristas que conectan a los vértices de un componente
			for (int i=indicesTotales.size()-1; i>=0; i--){
				int index = indicesTotales.get(i);
				componentes.get(ind).aristas.add(aristasVisitadas.get(index));
				aristasVisitadas.remove(index);
			}

			// Asigna la ruta de vertices generada para cada componente
			componentes.get(ind).vertices = ruta;
		}

		// Imprimir el contenido de cada componente generado
		for (int i=0; i<componentes.size(); i++){
			System.out.println("Componente: " + i);
			for (int j=0; j<componentes.get(i).vertices.size(); j++){
				System.out.print(componentes.get(i).vertices.get(j).nombre + ", ");
			}
			System.out.println("\n");
			for (int j=0; j<componentes.get(i).aristas.size(); j++){
				componentes.get(i).aristas.get(j).imprimirArista();
			}
			System.out.println(" ");
		}

		// Segunda y Final Iteración: conectar las componentes en una sola (árbol Mínimo)

		// Asignar las aristas minimas que conecten a cada nuevo componente
		for (int i=0; i<componentes.size(); i++){
			Comparator<Arista> comparador = new ComparadorArista(); // Ordena la cola de prioridad
			PriorityQueue<Arista> pq = new PriorityQueue<>(comparador);

			for (int j=0; j<componentes.get(i).vertices.size(); j++){
				for (int k=0; k<componentes.get(i).vertices.get(j).aristas.size(); k++){
					Arista act = componentes.get(i).vertices.get(j).aristas.get(k);

					// Si la arista adyacente al vertice no forma parte del componente, se agrega para
					// analizar conexiones entre componentes
					if (buscarArista(componentes.get(i).aristas, act) == false){
						pq.add(act);
					}
				}
			}

			// Una vez analizado todo el componente se procede a elegir la arista mínima
			Arista minima = pq.poll();

			// Verificar que las aristas no se dupliquen
			if (buscarArista(aristasVisitadas,minima) == false){
				aristasVisitadas.add(minima);
				aristasTotales.add(minima);
			}
		}

		return aristasTotales;
		
	}

	// Busca una arista almacenada en una lista (Listas de esta clase)
	public static boolean buscarArista(LinkedList<Arista> aristasVisitadas, Arista a){
		boolean encontrado = false;
		
		for (int i=0; i<aristasVisitadas.size(); i++){
			Arista aux = aristasVisitadas.get(i);
			
			if ((aux.v == a.v && aux.u == a.u) || (aux.v == a.u && aux.u == a.v)){
				encontrado = true;
			}
		}

		return encontrado;
	}

	// Busca una arista almacenada en una lista (Listas de la clase Grafo, Vertice, Arista)
	public static boolean buscarArista(List<Arista> aristasVisitadas, Arista a){
		boolean encontrado = false;
		
		for (int i=0; i<aristasVisitadas.size(); i++){
			Arista aux = aristasVisitadas.get(i);
			
			if ((aux.v == a.v && aux.u == a.u) || (aux.v == a.u && aux.u == a.v)){
				encontrado = true;
			}
		}

		return encontrado;
	}

	// Busca vertices conectados entre sí para añadirlos a un componente (hacia la izquierda)
	public static void buscarVerticeIzq(Vertice actual, LinkedList<Arista> aristasVisitadas, LinkedList<Vertice> ruta, List<Integer> indicesTotales){
		LinkedList<Integer> indices = new LinkedList<>();

		// Busca los vertices que conectan por la derecha
		for (int i=0; i<aristasVisitadas.size(); i++){
			if (aristasVisitadas.get(i).u == actual){
				indices.add(i);
				indicesTotales.add(i);
			}
		}

		// Añade los vertices a la ruta sin que se repitan
		for (int i=0; i<indices.size(); i++){
			Vertice aux = aristasVisitadas.get(indices.get(i)).v;
			añadirVertice(aux,ruta);
		}

		// Búsqueda recursiva
		for (int i=0; i<indices.size(); i++){
			buscarVerticeIzq(aristasVisitadas.get(indices.get(i)).v, aristasVisitadas, ruta, indicesTotales);
		}
	}

	// Busca vertices conectados entre sí para añadirlos a un componente (hacia la derecha)
	public static void buscarVerticeDer(Vertice actual, LinkedList<Arista> aristasVisitadas, LinkedList<Vertice> ruta, List<Integer> indicesTotales){
		LinkedList<Integer> indices = new LinkedList<>();

		// Busca los vertices que conectan por la izquierda
		for (int i=0; i<aristasVisitadas.size(); i++){
			if (aristasVisitadas.get(i).v == actual){
				indices.add(i);
				indicesTotales.add(i);
			}
		}

		// Añade los vertices a la ruta sin que se repitan
		for (int i=0; i<indices.size(); i++){
			Vertice aux = aristasVisitadas.get(indices.get(i)).u;
			añadirVertice(aux,ruta);
		}

		// Búsqueda recursiva
		for (int i=0; i<indices.size(); i++){
			buscarVerticeDer(aristasVisitadas.get(indices.get(i)).u, aristasVisitadas, ruta, indicesTotales);
		}
	}

	// Añade un vertice sin que se repita en una lista
	public static void añadirVertice(Vertice v,LinkedList<Vertice> ruta){
		boolean repetido = false;
		for (int i=0; i<ruta.size(); i++){
			if (ruta.get(i) == v){
				repetido = true;
			}
		}

		// Si no se repite el vertice, se inserta
		if (repetido == false){
			ruta.add(v);
		}
	}

	public static void bubbleSort(LinkedList<Integer> Lista){
		int n = Lista.size();
		for (int i=n-1; i>=0; i--){
			int k=0;
			for (int j=0; j<=i-1; j++) {
				if (Lista.get(j)>Lista.get(j+1)){
					int aux = Lista.get(j);
					Lista.set(j,Lista.get(j+1));
					Lista.set(j+1,aux);
				} else {
					k=k+1;
				}
			}
			if (k==i){
				break;
			}
		}
	}

}